File: repositories/ocr_repository.go
package repositories

import (
	"context"
	"errors"

	"doctor-service/entities"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type OCRRepository interface {
	Create(ctx context.Context, ocrResult *entities.DocumentOCRResult) error
	FindByDocumentID(ctx context.Context, documentID uuid.UUID) ([]entities.DocumentOCRResult, error)
	FindLatestByDocumentID(ctx context.Context, documentID uuid.UUID) (*entities.DocumentOCRResult, error)
}

type GormOCRRepository struct {
	db *gorm.DB
}

func NewOCRRepository(db *gorm.DB) OCRRepository {
	return &GormOCRRepository{db: db}
}

func (r *GormOCRRepository) Create(ctx context.Context, ocrResult *entities.DocumentOCRResult) error {
	return r.db.WithContext(ctx).Create(ocrResult).Error
}

func (r *GormOCRRepository) FindByDocumentID(ctx context.Context, documentID uuid.UUID) ([]entities.DocumentOCRResult, error) {
	var results []entities.DocumentOCRResult
	err := r.db.WithContext(ctx).Where("document_id = ?", documentID).Order("processed_at DESC").Find(&results).Error
	return results, err
}

func (r *GormOCRRepository) FindLatestByDocumentID(ctx context.Context, documentID uuid.UUID) (*entities.DocumentOCRResult, error) {
	var result entities.DocumentOCRResult
	err := r.db.WithContext(ctx).Where("document_id = ?", documentID).Order("processed_at DESC").First(&result).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &result, err
}

File: backend_api/repositories/ocr_repository.py
from typing import List, Optional
from uuid import UUID
from sqlalchemy.orm import Session
from entities.ocr_result import DocumentOCRResult

class OCRRepository:
    def __init__(self, db: Session):
        self.db = db

    def Create(self, ocr_result: DocumentOCRResult) -> None:
        self.db.add(ocr_result)
        self.db.commit()
        self.db.refresh(ocr_result)

    def FindByDocumentID(self, document_id: UUID) -> List[DocumentOCRResult]:
        return self.db.query(DocumentOCRResult).filter(DocumentOCRResult.document_id == document_id).order_by(DocumentOCRResult.processed_at.desc()).all()

    def FindLatestByDocumentID(self, document_id: UUID) -> Optional[DocumentOCRResult]:
        return self.db.query(DocumentOCRResult).filter(DocumentOCRResult.document_id == document_id).order_by(DocumentOCRResult.processed_at.desc()).first()
